src/APIBasedAuth.ts

```diff
@@ -0,0 +1,192 @@
+import axios, {
+  AxiosError,
+  AxiosInstance,
+  AxiosRequestConfig,
+  AxiosResponse
+} from 'axios';
+
+/* istanbul ignore next */
+const formatAxiosError = <ErrorResponse = any>(
+  error: AxiosError<ErrorResponse> | unknown
+) => {
+  if (!axios.isAxiosError(error)) {
+    throw new AxiosError(String(error));
+  }
+  throw error.response?.data;
+};
+
+class APIBasedAuth {
+  private client: AxiosInstance;
+  private clientOptions: APIBasedAuth.Config;
+  private session?: APIBasedAuth.Session;
+  private tokenData?: APIBasedAuth.TokenData;
+
+  constructor(options: APIBasedAuth.Config) {
+    const required = ['baseURL', 'clientId'];
+    for (const param of required) {
+      // eslint-disable-next-line security/detect-object-injection
+      if (!options[param]) {
+        throw new Error(`APIBasedAuth ctor param ${param} is required`);
+      }
+    }
+
+    this.client = axios.create({
+      baseURL: options.baseURL,
+      withCredentials: options.withCredentials
+    });
+    this.client.interceptors.request.use(this._setAuthorizationHeader);
+
+    this.clientOptions = options;
+  }
+
+  /* istanbul ignore next */
+  private async _setAuthorizationHeader(
+    request: AxiosRequestConfig<unknown>
+  ): Promise<AxiosRequestConfig<unknown>> {
+    if (this.tokenData?.access_token) {
+      request.headers = {
+        ...request?.headers,
+        Authorization: `Bearer ${this.tokenData.access_token}`
+      };
+    }
+    return request;
+  }
+
+  public confirmPasswordlessAuth(
+    baseInput: Omit<APIBasedAuth.VerifyPasswordlessAuthData, 'clientId'>
+  ) {
+    const input = {
+      ...baseInput,
+      // if session or username is not present get stored properties
+      ...(!baseInput.session || !baseInput.username ? this.session : null)
```

Comment:
Will `this.session` always have the up-to-date value from storage here?

Reply:
The motivation behind storing the session is primarily this flow:
1. Someone triggers passwordless.
2. They close their web page / kill their app.
3. They click on the link in their passwordless email, which re-opens their app.
4. Now, we need the `session` from step 1, which is no longer in memory (it's only in storage).

So, it's possible that `confirmPasswordlessAuth` might get called before `initiatePasswordlessAuth` is ever called (in a particular session). That makes me think we need to fetch from storage here.

Reply:
@virdesai did you have a thought about this? It seems like the flow that Swain pointed it is right, will a session exist at this point?

Reply:
Opps. I see that instead of fixing it here you opened a new PR. I'll review there